### DISCUSSION TITLE
Reproducibility in the long run

### DISCUSSION THREAD

**Discussion by rodluger** [rodluger/showyourwork#198] [General]
In #184, @adrn, @MilesCranmer, @mathren and I started chatting about things like the [longevity of GitHub links](https://github.com/showyourwork/showyourwork/pull/184#issuecomment-1250067186), the idea of archiving showyourwork repositories with [software heritage](https://github.com/showyourwork/showyourwork/pull/184#issuecomment-1250112417), and potentially relying on [Zenodo to archive source code](https://github.com/showyourwork/showyourwork/pull/184#issuecomment-1250401514) as well as data. I think this is all part of a much broader discussion concerning the future of showyourwork and reproducibility in astronomy in general. I suggest we continue the discussion in this thread, which I can pin to make more visible others.

My [two cents](https://github.com/showyourwork/showyourwork/pull/184#issuecomment-1251070862), just to get the conversation started, is that `showyourwork` is currently designed with short-term reproducibility in mind. We explicitly link to GitHub repositories (which can get deleted and have no real guarantee of true longevity), we manage dependencies with `conda` and `pip` (which rely on various package hosts and channels being around in the long run), and we store only the source code, not a full container of the workflow environment. This may well make it difficult, or not impossible, for someone to reproduce a paper built using `showyourwork` ten or twenty years from now. 

But should that really be our goal? To what extent should we strive to fully containerize articles, versus enable reproducibility in the short term? And how much of this is our responsibility, versus that of the journals we're submitting our papers to? How can we get journals to buy into the whole reproducibility thing?

And if we do want to move towards longer term reproducibility, how do we do that with `showyourwork`? Docker containers? Software heritage? Can we find a solution that is still practical in the short term?

**Comment by MilesCranmer** [comment]
Great points all around, I didn’t consider this angle. After reading your comment, I definitely agree that short-term reproducibility is by far the most important: both for (1) research verification, and (2) building off other’s work. Long-term reproducibility is arguably important for (2) in slower-moving research areas. 

But as you mention, long-term reproducibility is not only less important, it’s also **far** harder to pull off. (I would argue that storing code on GitHub is probably not even the weakest link in the reproducibility chain?) Each solution feels like it also has faults: docker requires compatibility of old images on future machines, software heritage assumes you can recursively bootstrap all the other dependencies (assuming conda isn’t around). 

I think the best we can reasonably do for the long term is just ensure the source files: the scripts + environment file + source TeX, are available indefinitely. arXiv is already used for one of these - how about automatically including the environment file + scripts in arXiv uploads? (Journals already don’t make TeX available, so I think arXiv is the only one that matters)

**Reply by rodluger** [reply]
[@MilesCranmer  replied](https://github.com/showyourwork/showyourwork/issues/191#issuecomment-1252952754): What if you could download an arXiv tar ball, unzip, and then re-build the figures with showyourwork (pulling data from zenodo)? I feel like that would be an elegant solution to the reproducibility question if GitHub is to be avoided. But would take some work to reconfigure everything. (arXiv seems like a good place to put source files - in addition to the build results which are typically uploaded)

**Reply by rodluger** [reply]
[@mathren replied](https://github.com/showyourwork/showyourwork/issues/191#issuecomment-1252966539): @MilesCranmer why are you thinking of using arxiv for this and not zenodo itself?

You could upload the code as well there together with the data. This is usually what I do for the MESA simulations in my publications, data output and a template input go on zenodo, so that as long as you can install MESA and type in the template the numbers I have used you can reproduce bit-to-bit my results. An example from a collaborator of mine is [here](https://zenodo.org/record/3559859) where the "reproducibility package" was done by hand. If showyourwork could automatize this it would be very helpful. I have anecdotal evidence that this works to guarantee reproducibility on a timescale of at least ~5 years (which is not very long, ok...)

On the long term, I think hosting code and output associated to papers should be something done by the journals themselves, but barring that it seems to me zenodo would be a better hosting for code than arxiv...

**Reply by rodluger** [reply]
[@MilesCranmer replied](https://github.com/showyourwork/showyourwork/issues/191#issuecomment-1252992398): Yes, you are right, zenodo is a much better choice than arXiv! Not sure what I was thinking. (I guess it might be most obvious for people if the other source files are with the TeX uploaded to arXiv - no linking needed.)

I think there is a GitHub action to regularly archive a repository on zenodo (code and all). Maybe that could be an option.

**Reply by rodluger** [reply]
Sorry for the editorializing, all! I want to try out this new Discussions board -- I think it will be a better way of organizing replies to specific posts within the discussion.

**Reply by rodluger** [reply]
I don't think shipping the source code with the manuscript on arXiv is a bad idea, especially since the goal here is to think of a "paper" as more than just the article PDF, but the entire workflow that sits behind it. Agreed that it would be nice to have it all in one place -- I think that is ultimately what we want out of refereed journals, maybe?

But for the time being I agree with @mathren that the Zenodo approach may be best. Not all papers get submitted to the arXiv, especially in other disciplines. Zenodo seems like a universal solution, at least for now.

**Reply by mkenworthy** [reply]
Resurrecting this year old thread, I'd argue that the most crucial element is the Snakemake file - that's the logical glue for all your work.

**Comment by arm61** [comment]
Just some thoughts from outside of the astronomy community (where reproducibility of analysis/paper/etc is also extremely important). The longer-term reproducibility goal will be very very hard, and potentially beyond the scope of something like `showyourwork`. The "long-term" reproducibility goal that @MilesCranmer suggests is both achievable and important. 

I have tried to achieve this in the past by making releases of my reproducible paper repositories (i.e. https://github.com/arm61/model_select/releases or https://github.com/arm61/lipids_at_airdes/releases) which trigger Zenodo to archive the state (resulting in https://zenodo.org/record/3697795#.Yyq0zC8Rpqs or https://zenodo.org/record/2577796#.Yyq04i8Rpqs respectively). I make these releases when I publish to ArXiv, after each round of peer review and finally when the paper is accepted/published. Along the way, I include a "Reproducibility Statement" in the text which references my Zenodo minted DOI. 

I am taking a similar approach when using `showyourwork` in my current work (https://github.com/arm61/reporting_sampling) although I am slightly changing the vocabulary I use. In general, using `showyourwork` enables better accountability of my reproducibility rather than a bunch of hand-rolled (Snake)Make files. 

A final thought on this point from @MilesCranmer:

> I think there is a GitHub action to regularly archive a repository on zenodo (code and all). Maybe that could be an option.

I would say this is undesirable as you want to give the user control rather than it happening automatically. I don't necessarily want a Zenodo archive when I have dealt with half the reviewers' comments, I want it at the point of resubmission. 

TL;DR. My suggestion is to use the GitHub releases and Zenodo integration (and give documentation on how to do this) to enable this "long-term" reproducibility.

**Reply by rodluger** [reply]
Great points! Yes, I've used the Zenodo hook to create a deposit on new releases, and it's a great idea. Agreed it should happen on the user's schedule. I think, though, it might be nice to run an action whenever this happens that records the most recent Zenodo DOI in the article itself -- so we can actually link directly to the deposit from the paper.

**Reply by arm61** [reply]
The problem I found (and still find with this) is that I write my paper and make a release at a given commit, this triggers a Zenodo deposit and produces a DOI, to then include this DOI in my paper I have to create a new commit on the repo...which kinda means that the repo and the Zenodo deposit no longer match. 

I don't think that there is a way out of the vicious cycle (but an action could make it more accountable ensuring that the only change that has happened is the addition of the Zenodo DOI).

**Reply by rodluger** [reply]
Yeah that's a classic one. I don't think it's a big deal, though. There is the option, however, to pre-reserve a DOI on Zenodo, so we can link to it in advance of publishing the deposit. I believe we would have to write our own hook to upload the deposit in this case.

**Reply by arm61** [reply]
Yeah the pre-reserving (not a Zenodo but a (non-versioning) institutional repository) was an approach I took for data in the past. Would be cool, but writing your own hook might not be worth it.
